### PR TITLE
fix: prioritize special rules in manifest files while checking app dependencies

### PR DESCRIPTION
**Pull request by hfudev**
also update the documents for config rules

### CODE REVIEW

File: docs/manifest.md [hunk 1]
```diff
@@ -128,17 +128,23 @@ To enable this feature, you need to declare the modified files or components usi
 - `--modified-files`
 - `--modified-components`
 
-To declare the dependencies for an app, you could add two more keywords in the corresponding manifest file:
+By default, one app will be built when:
```
Comment (ESP-Marius) [inline docs/manifest.md:131]:
```suggestion
By default, the app will be built when:
```

File: docs/manifest.md [hunk 2]
```diff
@@ -156,17 +162,16 @@ This app will be built with the following CLI options:
...
 
 - `--modified-files examples/foo/main/foo.md`
 - `--modified-components bar`
-- `/tmp/foo.c`
 
-### Advanced Usage
+### Ignore Root Dependencies
 
-Sometimes, we have some root dependencies. All apps should be built if these root dependencies are modified.
+Sometimes, we have some root dependencies that is really low-level. All apps should be built if these root dependencies are modified. We can use the `--ignore-root-dependencies` CLI option to achieve this.
```
Comment (ESP-Marius) [inline docs/manifest.md:175]: Maybe need more details here? What are the root dependencies? Set by the users? A default selection? What exactly am I skipping when using `---ignore-root-dependencies`?

File: docs/manifest.md [hunk 3]
```diff
@@ -128,17 +128,23 @@ To enable this feature, you need to declare the modified files or components usi
...
 
-One app will be built when:
+When any of these two entries are specified, the default behavior will be different. Instead, the app will be built when:
 
 - The app itself is modified (`.md` files are excluded)
-- Any of the specified `depends_components` in the corresponding manifest file are modified
-- Files that matches any of the specified `depends_filepatterns` in the corresponding manifest file are modified
-- Any of the `build_components` are modified. `build_components` are defined in the `project_description.json`, which is generated by running `idf.py reconfigure` or `idf.py build`
+- Any of the specified `depends_components` is modified
+- Any file that matches any of the specified `depends_filepatterns` is modified
```
Comment (ESP-Marius) [inline docs/manifest.md:148]: Should we write it explicitly that when using these options `build_components` no longer affect the build? Just to make it 100% clear.

File: docs/manifest.md [hunk 4]
```diff
@@ -128,17 +128,23 @@ To enable this feature, you need to declare the modified files or components usi
 - `--modified-files`
 - `--modified-components`
 
-To declare the dependencies for an app, you could add two more keywords in the corresponding manifest file:
+By default, one app will be built when:
+
+- The app itself is modified (`.md` files are excluded)
+- Any of the `build_components` are modified.
+
+    `build_components` are defined in the `project_description.json`, which is generated by running `idf.py reconfigure` or `idf.py build`
```
Comment (igrr) [inline docs/manifest.md:136]:
```suggestion
- Any of the components used by the app are modified.

    The list of components used by the app is obtained from `BUILD_COMPONENTS` property in IDF build system. `idf-build-apps` runs `idf.py reconfigure` to determine the value of this property for the given app.
```

File: docs/manifest.md (hunk 3 above)
Comment (igrr) [reply docs/manifest.md:148]: I think we can make this clear by restructuring the description a bit:

---
## Building apps only on related changes

In large projects or monorepos it is often desirable to only run builds and tests which are somehow related to the changes in a pull request.

`idf-build-apps`  supports this by checking whether a particular app has been modified, or depends on modified components or modified files.

### Basic Usage

To enable this feature, you need to pass the list of modified files or modified components to idf-build-apps using the following CLI options:

- `--modified-files`
- `--modified-components`

For example, if the project uses Git, you can obtain the list of files modified in a branch or a pull request by calling
```bash
git diff --name-only ${pr_branch_head} $(git merge-base ${pr_branch_head} main)
```
where `pr_branch_head` is the branch of the pull request, and `main` is the default branch.

idf_build_apps uses the following rules to determine whether to build an app or not:

1. Any files in the app itself are modified (`.md` files are excluded)
2. If `depends_components` or `depends_filepatterns` are specified in the manifest file, idf-build-apps matches `modified-components` and `modified-files` against these two options. If any of the modified components are in the `depends_components` list or any of the modified files are matched by `depends_filepatterns`, the app is built.
3. If `depends_components` or `depends_filepatterns` are not specified in the manifest file, idf-build-apps determines the list of components the app depends on using `BUILD_COMPONENTS` property in IDF build system. For the given app, this property contains the list of all the components included into the build. idf_build_apps runs `idf.py reconfigure` to determine the value of this property for the app. If any of the modified components are present in BUILD_COMPONENTS list, the app is built.

File: docs/manifest.md (hunk 2 above)
Comment (igrr) [reply docs/manifest.md:175]: I'm also a bit confused, 'ignore-root-dependencies' is used to achieve 'apps should be built if these root dependencies are modified'? Not the other way around?

**Comment by hfudev** [comment]
@igrr @ESP-Marius Thank you for your review! All your suggestions are applied. I also rewrite the last paragraph. PTAL again.

The test coverage should be fine. The `find_apps` function basically calls the `check_should_build` function, but set a few more variables, like manifest files. I modified the test scripts, to make the files mentioned in the test function actually exists, to be matched. (I think this is a bug before) I'll double-check the coverage once the pipeline is passing.

**Pull request merged**